src/object.hpp

```diff
@@ -178,20 +191,323 @@ template <typename Derived> class readable_object {
         return as<T>();
     }
 
+    // is<T> checks whether the underlying type is compatible with the required
+    // type. When it comes to numeric types, the request type must match the
+    // one used within ddwaf_object, i.e. the type will not be cast to one of
+    // a smaller size.
+    template <typename T>
+    [[nodiscard]] bool is() const noexcept
+        requires is_type_in_set_v<T, bool, int64_t, uint64_t, double, std::string, std::string_view,
+            const char *>
+    {
+        return is_compatible_type<T>(type());
+    }
+
+    // Overload for other unsigned integer types
+    template <typename T>
+    [[nodiscard]] bool is() const noexcept
+        requires(!std::is_same_v<T, uint64_t>) && std::is_integral_v<T> && std::is_unsigned_v<T> &&
+                (!std::is_same_v<T, bool>)
+    {
+        using limits = std::numeric_limits<T>;
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return is_compatible_type<uint64_t>(type()) && obj.uintValue <= limits::max();
+    }
+
+    // Overload for other signed integer types
+    template <typename T>
+    [[nodiscard]] bool is() const noexcept
+        requires(!std::is_same_v<T, int64_t>) && std::is_integral_v<T> && std::is_signed_v<T>
+    {
+        using limits = std::numeric_limits<T>;
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return is_compatible_type<int64_t>(type()) && obj.intValue >= limits::min() &&
+               obj.intValue <= limits::max();
+    }
+
+    // Convert the underlying type to the requested type, converters are defined
+    // in the object_converter header
```

Comment:
that header should then be included in this file (and ideally object_converter.hpp be made a private header)